towntest_chest/init.lua

```diff
@@ -158,16 +208,23 @@ towntest_chest.build = function(chestpos)
 	for i,v in ipairs(building_plan) do
 		local pos = {x=v.x+chestpos.x,y=v.y+chestpos.y,z=v.z+chestpos.z}
 		-- check if the builder has the node
-		if inv:contains_item("builder", v.name) then
+		if inv:contains_item("builder", v.matname) or -- is payed or
+-- for free and the item is at the end of building plan (all next items already built, to avoid all free items are placed at the first)
+		   ( v.matname == "free" and i == #building_plan )
+		then
 			-- check if npc is already moving
 			if npc and not npc.target then
 				table.remove(building_plan,i)
 				-- move the npc to the build area
 				npc:moveto({x=pos.x, y=pos.y+1.5, z=pos.z}, 2, 2, 0, function(self,after_param)
 					-- take from the inv
-					after_param.inv:remove_item("builder", after_param.v.name.." 1")
+					after_param.inv:remove_item("builder", after_param.v.matname.." 1")
 					-- add the node to the world
 					minetest.env:add_node(after_param.pos, {name=after_param.v.name,param1=after_param.v.param1,param2=after_param.v.param2})
+			                if v.meta then
+						minetest.env:get_meta(after_param.pos):from_table(after_param.v.meta)
```

Comment:
It was hard to understand for me too, the line is copied from worldedit restoring nodes implementation. Seems the "get_meta" returns the reference, so the from_table changes directly the data in minetest-node.
There is a bug that I fixed yesterday.